Add External Storage Backend Support

happyhackerhour opened the pull request:
# Add External Storage Backend Support

## Summary

This PR adds support for external storage backends to snap-cinder-volume, enabling integration with enterprise storage arrays from Hitachi, Pure Storage, and Dell EMC.

## New Storage Backends

### Hitachi VSP Storage
- **Protocols**: Fibre Channel and iSCSI
- **Configuration**: `hitachi.<backend-name>.*` namespace
- **Driver**: Hitachi HBSD FC/iSCSI drivers

### Pure Storage FlashArray  
- **Protocols**: iSCSI, Fibre Channel, and NVMe
- **Configuration**: `pure.<backend-name>.*` namespace
- **Driver**: Pure Storage FC/iSCSI/NVMe drivers

### Dell Storage Center
- **Protocols**: iSCSI and Fibre Channel
- **Configuration**: `dellsc.<backend-name>.*` namespace
- **Driver**: Dell EMC Storage Center drivers

## Configuration Examples

**Hitachi VSP:**
```bash
sudo snap set cinder-volume \
  hitachi.vsp350.volume-backend-name=vsp350 \
  hitachi.vsp350.san-ip=10.0.0.50 \
  hitachi.vsp350.san-login=admin \
  hitachi.vsp350.san-password=password \
  hitachi.vsp350.hitachi-storage-id=45000 \
  hitachi.vsp350.hitachi-pools=DP_POOL_01
```

**Pure Storage:**
```bash
sudo snap set cinder-volume \
  pure.flasharray1.volume-backend-name=flasharray1 \
  pure.flasharray1.san-ip=10.0.0.100 \
  pure.flasharray1.pure-api-token=your-token \
  pure.flasharray1.protocol=iscsi
```

**Dell Storage Center:**
```bash
sudo snap set cinder-volume \
  dellsc.sc4020.volume-backend-name=sc4020 \
  dellsc.sc4020.san-ip=10.0.0.200 \
  dellsc.sc4020.san-login=admin \
  dellsc.sc4020.san-password=password \
  dellsc.sc4020.dell-sc-ssn=64702
```

## Implementation

- **Dynamic backend discovery** - Automatically detects and configures new backend types
- **Generic template system** - Single template handles all backend configurations  
- **Automatic field conversion** - Handles kebab-case to snake_case conversion for OpenStack compatibility
- **Comprehensive validation** - Ensures unique backend names and valid configurations

This enables snap-cinder-volume to work with enterprise storage infrastructure beyond just Ceph.

gboutry commented on cinder_volume/configuration.py:
```diff
@@ -91,6 +105,202 @@ class CephConfiguration(BaseBackendConfiguration):
     rbd_secret_uuid: str
     rbd_key: str
 
+class HitachiConfiguration(BaseBackendConfiguration):
```
Are all this configuration option really necessary?

It feels like we're redefining defaults?

I wonder if we should just handle extra dict config automatically?

gboutry replied:
I wonder if we couldn't just do some extract logic as to, when generating context, to put all unknown fields into an `extra_fields` in the context dict, and just render that for unknown config.

This would allow users to manually override on the now when there is an issue, while we'll be able to focus on the mandatory options, well managing them. What do you think?

happyhackerhour replied:
you are totatlly rigth, I think it makes more sense to support all backends in a generic way and delegate configuration validation and driver selection to the charms. That leaves us with a simpler and easier-to-maintain design. I’ll update the code to use a generic backend that accepts any options, along with a generic template that renders a per-backend file in cinder.conf.d.

gboutry replied:
that's not exactly what I meant.

I think we should still manage purposeful config options for each backend.

Just not ALL of them, while still allowing for random config options to be set.

happyhackerhour replied:
Done

happyhackerhour replied:
Oh! just read your comment. ok. i have done a generic implementaion. 

let me now if you want it changed to a per vendor implementation, with only the requied options and the rest of them removed.

i think this proposed generic way is easier to mantain, and transfers the repsonsabliity upstream to set the correct values.  but i can change it to a per vendor implementation with just the minimal required values. with allowed extra options. 

let me know!

happyhackerhour replied:
simplified the config models with only the required fields for each backend.
add support for extra fiedls (not specified in the pydantic model) to be rendered in the final configuration.

gboutry commented on cinder_volume/configuration.py:
```diff
@@ -18,9 +18,23 @@
+from pydantic import Field, ValidationInfo, field_validator
 import pydantic.alias_generators
 
+__all__ = [
+    "DatabaseConfiguration",
+    "RabbitMQConfiguration",
+    "CinderConfiguration",
+    "CephConfiguration",
+    "HitachiConfiguration",
+    "PureConfiguration",
+    "DellSCConfiguration",
+    "Configuration",
+]
```
Please do not restrain exposed values.

gboutry commented on cinder_volume/configuration.py:
```diff
@@ -18,9 +18,23 @@
 takes as input from `snap set`.
 """
 
+from __future__ import annotations
```
How is that needed?

The charm is running in python3.12 or later. I don't think we need that ?

gboutry commented on cinder_volume/cinder_volume.py:
```diff
@@ -254,24 +267,66 @@ def template(self, snap: Snap) -> list[template.Template]:
+            backend_ctxs: dict[str, context.Context] = {}
+
+            # Ceph back-ends
+            for name, be_cfg in cfg.ceph.items():
```
Not to be fixed now, but we might need to think about a smarter way to handle all these backends in the future...

happyhackerhour replied:
Implemented this in a dynamic way :)

gboutry reviewed (changes requested):
This is overall a very good work.

I think we'll need to be smarter on how we manage the options not to rewrite the world for each backend. I've suggested some way to do it.

We'll need to setup the proper ci workflow to validate the minimal aspect such as linting and building...

happyhackerhour commented:
hey Guillaume, thanks very much for your valuable time... yesterday was preaty busy. but today i squezed some time and fixed the reviews.

i have changed the snap-cinder-volume code. to handle generic backends(did not found any way to support every posible option). so rigth now is restricted to the specified backeds:

    #Supported backend namespaces - add new backend types here
    These are the top-level configuration keys that will be treated as backend configurations
    BACKEND_KEYS = {
    'hitachi', # Hitachi VSP storage arrays
    'pure', # Pure Storage FlashArray
    'dellsc', # Dell Storage Center (Compellent)
    'test', # Test backend for validation
    # Add new backend types here as they are supported
    }
following your recomendations, this will make code easier to mantain and add support for any backend.

the driver selection, and all the logic should be handled by the charm
